### ISSUE TITLE
Advance the Programming Model for Application Initialization at Build Time

### ISSUE THREAD

**Issue by christianwimmer** [oracle/graal#4922]
### TL;DR

The existing `Feature` mechanism allows applications to be partly initialized at image build time. That is usually used together with the option `--initialize-at-build-time` to initialize classes at image build time. We want to advance the programming model so that fewer classes need to be manually marked as "initialize at build time", and so that it is easier to run individual methods at image build time.

### Goals

* Reduce the need to use `--initialize-at-build-time`: Only classes that have instances on the image heap need to be explicitly marked as "initialize at build time".
* Provide a registration mechanism (probably annotation-based at first) to execute individual methods at image build time.

### Non-Goals

* The changes will be neutral regarding peak performance and startup performance. There should be no regressions, but there will be also no improvements until users start to leverage the new mechanism.
* There should be no compatibility issues: applications that are properly configured for initialization at build time / run time should continue to work.

### Details

Details about the new class initialization strategy, including a prototype implementation, are available here: #4684
